presto-main/src/main/java/com/facebook/presto/execution/scheduler/SourcePartitionedScheduler.java

```diff
@@ -340,28 +344,38 @@ else if (pendingSplits.isEmpty()) {
             return new ScheduleResult(false, overallNewTasks.build(), overallSplitAssignmentCount);
         }
 
-        if (anyBlockedOnPlacements) {
+        ScheduleResult.BlockedReason blockedReason;
+        if (anyBlockedOnNextSplitBatch) {
+            blockedReason = anyBlockedOnPlacements ? MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE : WAITING_FOR_SOURCE;
+        }
+        else {
+            blockedReason = anyBlockedOnPlacements ? SPLIT_QUEUES_FULL : NO_ACTIVE_DRIVER_GROUP;
+        }
+
+        if (blockedReason == MIXED_SPLIT_QUEUES_FULL_AND_WAITING_FOR_SOURCE
+                || blockedReason == SPLIT_QUEUES_FULL
+                || (groupedExecution && blockedReason == NO_ACTIVE_DRIVER_GROUP)) {
             // In a broadcast join, output buffers of the tasks in build source stage have to
             // hold onto all data produced before probe side task scheduling finishes,
             // even if the data is acknowledged by all known consumers. This is because
             // new consumers may be added until the probe side task scheduling finishes.
             //
-            // As a result, the following line is necessary to prevent deadlock
-            // due to neither build nor probe can make any progress.
+            // As a result, all tasks are proactively created when schedule is
+            // blocked due to SPLIT_QUEUES_FULL or NO_ACTIVE_DRIVER_GROUP
+            // to prevent deadlock.
+            // The deadlock is caused by neither build nor probe can make any progress.
```

Comment:
@dain  : The next two sentences aim at explain why this could happen. I will make it more explicit (by adding `as explained follows and put next two sentences into bullet points). Thank you !